### PR TITLE
DRAFT: Adds zero123

### PR DESCRIPTION
**Pull request by voletiv**
Makes it work with zero123.

`python launch.py --config configs/zero123.yaml --train --gpu 0`

TODOs
- Show 1 good result with correct lambda_rgb, etc.
- Update README
- Test with multiple GPUs
- Add acknowledgements to stable-dreamfusion for zero123 code

### CODE REVIEW

**Comment by voletiv** [comment]
Yup, I've put "Update README" in the TODOs above.
For the error you posted, just download the zero123 ckpt, I've written it in `load/zero123/download.sh`

**Comment by claforte** [comment]
One more TODO: we should acknowledge that the zero123 (and many of the upcoming ideas) are copied or adapted from Stable-DreamFusion

**Comment by voletiv** [comment]
Results so far with hamburger at 1000 iters

https://github.com/threestudio-project/threestudio/assets/22424247/4048834b-d484-4ff1-a45e-437047dfbdaf

File: configs/zero123.yaml [hunk 1]
```diff
@@ -0,0 +1,143 @@
+name: "zero123"
+tag: "${rmspace:${system.prompt_processor.prompt},_}"
```
Comment (thuliu-yt16) [inline configs/zero123.yaml:2]: Consider replacing the tag with the data image path since there is no prompt for this exp, i.e. `${rmspace:${basename:${data.image_path}},_}"` but not checked. See `threestudio/utils/config.py` for the custom function.

File: requirements.txt [hunk 2]
```diff
@@ -15,7 +15,7 @@ imageio[ffmpeg]
 trimesh
 git+https://github.com/NVlabs/nvdiffrast.git
 # deepfloyd
-git+https://github.com/facebookresearch/xformers.git@main#egg=xformers
+xformers
```
Comment (thuliu-yt16) [inline requirements.txt:18]: Why this change?

File: threestudio/models/renderers/nerf_volume_renderer.py [hunk 3]
```diff
@@ -176,6 +176,27 @@ def forward(
...
+                )
+                comp_normal = F.normalize(comp_normal, dim=-1)
+                comp_normal = (comp_normal + 1.0) / 2.0 * opacity  # for visualization
+                out.update(
+                    {
+                        "comp_normal": comp_normal.view(batch_size, height, width, 3),
+                    }
+                )
+                # if self.material.cfg.lambda_3d_normal_smooth > 0: # TODO
+                normal_perturb = self.geometry(
```
Comment (thuliu-yt16) [inline threestudio/models/renderers/nerf_volume_renderer.py:195]: consider adding an option to control to perturb or not since this will also introduce extra useless inference for other methods

File: threestudio/systems/zero123.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+            else:
+                bg_color = torch.rand(3).to(self.device)
+            ambient_ratio = 0.1 + 0.9 * random.random()
+        else:
+            # bg_color = torch.rand_like(batch['rays_o'])
+            ambient_ratio = 1.0
+            shading = "diffuse"
+            batch["shading"] = shading
+            bg_color = None
+
```
Comment (thuliu-yt16) [inline threestudio/systems/zero123.py:71]: Consider beginning with `if do ref` to be consistent with the logic below

File: threestudio/models/guidance/zero123_guidance.py [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+            .to(self.device)
+        )
+        self.c_crossattn, self.c_concat = self.get_img_embeds(self.rgb_256)
+
+    @torch.cuda.amp.autocast(enabled=False)
+    @torch.no_grad()
+    def get_img_embeds(
+        self,
+        img: Float[Tensor, "B 3 256 256"],
+    ):
```
Comment (thuliu-yt16) [inline threestudio/models/guidance/zero123_guidance.py:171]: Add type hint for c and v here and would you mind replacing `c, v` with some more informative names?

File: threestudio/models/guidance/zero123_guidance.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+    # verification - requires `vram_O = False` in load_model_from_config
+    @torch.no_grad()
+    def gen_from_cond(
+        self,
+        cond,
+        ddim_steps=50,
+        scale=3,
+        post_process=True,
+    ):
+        # produce latents loop
```
Comment (thuliu-yt16) [inline threestudio/models/guidance/zero123_guidance.py:402]: What is the difference between gen_from_cond and generate? It seems that they both use c_crossattn.

**Review by thuliu-yt16 — CHANGES_REQUESTED** [review]
I have some comments and questions about the code. Please check and resolve them. And I think it's better to put ldm_zero123 in a different folder, like a new `extern` folder instead of the root, and later we could turn such module into a git submodule. Please also resolve the conflict.

Thank you!

File: threestudio/models/renderers/nerf_volume_renderer.py (hunk 3 above)
Comment (bennyguo) [reply threestudio/models/renderers/nerf_volume_renderer.py:195]: I think a temporal solution could be adding options to the renderer. `self.cfg.return_comp_normal` and `self.cfg.return_normal_perturb` sound fine to me.

**Comment by bennyguo** [comment]
By the way the logic for guidance has changed in [this PR](https://github.com/threestudio-project/threestudio/pull/77). Please modify the zero123-guidance and the system accordingly. The changes are minor, and you may refer to other guidance and systems for examples.

File: requirements.txt (hunk 2 above)
Comment (claforte) [reply requirements.txt:18]: IIRC `git+https://github.com/facebookresearch/xformers.git@main#egg=xformers` wouldn't install on our cluster due to an incompatibility between the particular pytorch and CUDA toolkit versions, while `xformers` installed fine.

Comment (bennyguo) [reply requirements.txt:18]: However as indicated in the [xformers README](https://github.com/facebookresearch/xformers#installing-xformers), installing directly from `pip` requires `torch>=2.0.0`. Which PyTorch version are you using?

Comment (claforte) [reply requirements.txt:18]: Personally I'm using torch 2.x but IIRC, I previously tried installing pytorch 1.x with your instructions, and that's when I bumped into that xformers incompatibility. @voletiv can you check on your side if you can reproduce the problem with 1.x? Otherwise we can revert that particular change.

File: threestudio/models/renderers/nerf_volume_renderer.py (hunk 3 above)
Comment (voletiv) [reply threestudio/models/renderers/nerf_volume_renderer.py:195]: I reused `lambda_normal_smooth` and `lambda_3d_normal_smooth` from the yaml file

File: threestudio/models/guidance/zero123_guidance.py (hunk 6 above)
Comment (voletiv) [reply threestudio/models/guidance/zero123_guidance.py:402]: `generate` generates from `elevation`, `azimuth`, `camera_distance`.
`gen_from_cond` generates from `cond`.
I've simplified `generate` to call `gen_from_cond`.

File: requirements.txt (hunk 2 above)
Comment (voletiv) [reply requirements.txt:18]: I couldn't install tinycudann itself in torch 1.12, so the code doesn't run with 1.x in my case.

Comment (voletiv) [reply requirements.txt:18]: Screenshot 2023-05-29 at 16 26 05

**Comment by voletiv** [comment]
I've resolved the comments in the latest commits!
Regarding xformers, I was also able to install it in torch 2.x, I'm using '2.0.1+cu118'

**Comment by voletiv** [comment]
> I have some comments and questions about the code. Please check and resolve them. And I think it's better to put ldm_zero123 in a different folder, like a new `extern` folder instead of the root, and later we could turn such module into a git submodule. Please also resolve the conflict.
> 
> Thank you!

Done! I've put `ldm_zero` inside `extern`

**Comment by claforte** [comment]
@bennyguo @thuliu-yt16 I'd appreciate if you could merge this PR during our work night (even if it may still require changes) as I have a lot of further changes that depend on it, and for which I plan to submit PRs tomorrow during my work day. Thank you. BTW if you have further suggestions for Vikram and I please rest assured we'll address them soon after the PR is merged. Thanks!

**Comment by claforte** [comment]
BTW the TODOs (e.g. update the README) remain outstanding, but IMHO this shouldn't prevent merging this PR tonight - we can do it tomorrow.

**Review by bennyguo — APPROVED** [review]

**Review by bennyguo — APPROVED** [review]
